src/gfn/examples/distributions.py

```diff
@@ -0,0 +1,221 @@
+    are Mixture of Beta distributions on the possible angle range.
+
+    Adapted from https://github.com/saleml/continuous-gfn/blob/master/sampling.py
+
+    This is useful for the `Box` environment
+    """
+
+    def __init__(
+        self,
+        delta: float,
+        northeastern: bool,
+        centers: TensorType["n_states", 2],
+        mixture_logits: TensorType["n_states", "n_components"],
+        alpha: TensorType["n_states", "n_components"],
+        beta: TensorType["n_states", "n_components"],
+    ):
+        self.delta = delta
+        self.northeastern = northeastern
+        self.centers = centers
+        self.n_states = centers.shape[0]
+        self.n_components = mixture_logits.shape[1]
+
+        assert mixture_logits.shape == (self.n_states, self.n_components)
+        assert alpha.shape == (self.n_states, self.n_components)
+        assert beta.shape == (self.n_states, self.n_components)
+
+        self.base_dist = MixtureSameFamily(
+            Categorical(logits=mixture_logits),
+            Beta(alpha, beta),
+        )
+
+        self.min_angles, self.max_angles = self.get_min_and_max_angles()
+
+    def get_min_and_max_angles(self):
+        if self.northeastern:
+            min_angles = torch.where(
+                self.centers[:, 0] <= 1 - self.delta,
+                0.0,
+                2.0 / torch.pi * torch.arccos((1 - self.centers[:, 0]) / self.delta),
+            )
+            max_angles = torch.where(
+                self.centers[:, 1] <= 1 - self.delta,
+                1.0,
+                2.0 / torch.pi * torch.arcsin((1 - self.centers[:, 1]) / self.delta),
+            )
+        else:
+            min_angles = torch.where(
+                self.centers[:, 0] >= self.delta,
+                0.0,
+                2.0 / torch.pi * torch.arccos((self.centers[:, 0]) / self.delta),
+            )
+            max_angles = torch.where(
+                self.centers[:, 1] >= self.delta,
+                1.0,
+                2.0 / torch.pi * torch.arcsin((self.centers[:, 1]) / self.delta),
+            )
+
+        return min_angles, max_angles
+
+    def rsample(self, sample_shape=()):
+        base_01_samples = self.base_dist.rsample(sample_shape=sample_shape)
+
+        sampled_angles = (
+            self.min_angles + (self.max_angles - self.min_angles) * base_01_samples
+        )
+        sampled_angles = torch.pi / 2 * sampled_angles
+
+        sampled_actions = (
+            torch.stack(
+                [torch.cos(sampled_angles), torch.sin(sampled_angles)],
+                dim=1,
+            )
+            * self.delta
+        )
+
+        if self.northeastern:
+            sampled_next_states = self.centers + sampled_actions
+        else:
+            sampled_next_states = self.centers - sampled_actions
+        # TODO: do we need to return actions or next states here? -- logprobs are ok, but the previous 4 lines might not be
+        return sampled_next_states
```

Comment:
I don't fully understand this TODO -- my immediate thought if that `sampled_next_states`  should *always* be the only thing we return ?

Reply:
Actions are useful when printing trajectories. But it probably is their only benefit. If we decide we do not care about this, we could just have action samplers sample next states, and discard actions altogether.